lib/smart_proxy_dns_powerdns/backend/rest.rb

```diff
@@ -0,0 +1,92 @@
+require 'json'
+require 'net/http'
+require 'resolv'
+
+module Proxy::Dns::Powerdns::Backend
+  class Rest < ::Proxy::Dns::Powerdns::Record
+
+    attr_reader :url, :api_key
+
+    def initialize(a_server = nil, a_ttl = nil)
+      @url = Proxy::Dns::Powerdns::Plugin.settings.powerdns_rest_url
+      @api_key = Proxy::Dns::Powerdns::Plugin.settings.powerdns_rest_api_key
+
+      super(a_server, a_ttl)
+    end
+
+    def get_zone name
+      fqdn = Resolv::DNS::Name.create(name)
+      uri = URI("#{url}/zones")
+
+      domain = nil
+
+      Net::HTTP.start(uri.host, uri.port) do |http|
+        request = Net::HTTP::Get.new uri
+        request['X-API-Key'] = @api_key
+        response = http.request request
+        zones = JSON.parse(response.body) rescue []
+
+        domain = zones.select { |zone|
+          domain = Resolv::DNS::Name.create(zone['name'])
+          domain == fqdn or fqdn.subdomain_of?(domain)
+        }.max_by { |zone| zone['name'].length }
+      end
+
+      raise Proxy::Dns::Error, "Unable to determine zone. Zone must exist in PowerDNS." unless domain
+
+      domain
+    end
+
+    def create_record domain_id, name, type, content
+      patch_records domain_id, name, type, content
+    end
+
+    def delete_record domain_id, name, type
+      patch_records domain_id, name, type, nil
+    end
+
+    def patch_records domain_id, name, type, content
+      uri = URI("#{url}/zones/#{domain_id}")
+
+      if content
```

Comment:
I'd suggest having dedicated methods for 'replace' and 'delete' 'changetype's instead of using a conditional. You can extract the common bits into a method though.